### ISSUE TITLE
Import useful error messages from silica

### ISSUE THREAD

**Issue by benjaminweb** [well-typed/optics#199]
Does it make sense to incorporate the error messages from https://github.com/mrkgnao/silica?

**Comment by adamgundry** [comment]
Right, I've been meaning to do a comparison of the messages in some typical error cases between `lens`, `silica` and `optics`. The hard part here is figuring out what the "typical error cases" actually are. Perhaps some inspiration can be gained from https://www.reddit.com/r/haskell/comments/69ots1/notorious_lens_errors_thread/dh9r4pi/

**Comment by adamgundry** [comment]
@kcsongor pointed out some tricks that `silica` uses and we could copy:
 * extend the `TypeError`s in the failing `Is` superclass and in `Join` with more informative error messages
 * pretty-print `OpticKind` using a type family to drop the `A_` prefix
 * add instances for classes like `Field1` using `TypeError` to explain that the type does not have enough fields.

**Comment by arybczak** [comment]
> pretty-print OpticKind using a type family to drop the A_ prefix

I like these prefixes, error messages read like English with them.
